File: cron/bankd/parse_deposits.php
<?php
require '../../util.php';

function b_query($query)
{
    #echo $query;
    do_query($query);
}

function deposref_exists($deposref)
{
    $query = "
        SELECT 1
        FROM
            users
        WHERE
            deposref='$deposref'
        LIMIT 1
        ";
    $result = do_query($query);
    return has_results($result);
}

$query = "
    SELECT
        bid, entry
    FROM
        bank_statement
    WHERE
        reqid IS NULL
        AND status='VERIFY'
    ";
$result = do_query($query);

while ($row = mysql_fetch_array($result)) {
    $bid = $row[0];
    $query = "
        UPDATE
            bank_statement
        SET
            status='PROC'
        WHERE
            bid='$bid'
            AND status='VERIFY'
        ";
    b_query($query);
    $line = $row[1];
    print "$line\n";
    $info = split(',', $line);
    if ($info[5] != '') {
        echo "Skipping payment out...\n";
        continue;
    }
    $acc = split('[.]', $info[4]);
    if (count($acc) < 2) 
        $deposref = $info[4];
    else 
        $deposref = $acc[1];

    $deposref = trim($deposref, " \"'\n");
    if (!deposref_exists($deposref)) {
        echo "\n$deposref deposref doesn't exist!\n\n";
        exit(-1);
    }
    $amount = $info[6];
    $amount = numstr_to_internal($amount);
    print "$deposref <= $amount\n";

    $query = "
        INSERT INTO requests (
            req_type,
            uid,
            amount,
            curr_type
        ) SELECT
            'DEPOS',
            uid,
            '$amount',
            'GBP'
        FROM users
        WHERE
            deposref='$deposref'
        ";
    b_query($query);
    $query = "
        UPDATE
            bank_statement
        SET
            reqid=LAST_INSERT_ID(),
            status='FINAL'
        WHERE
            bid='$bid'
        ";
    b_query($query);
}

File: cron/bankd/parse_deposits.py
import MySQLdb
import decimal

import re

db_password = ""

db = MySQLdb.connect("localhost","intersango",db_password,"intersango")

c = db.cursor()
c.execute("SELECT bid,entry FROM bank_statement WHERE status='VERIFY' and reqid IS NULL")
for bid,entry in c:
    try:
        line = entry.split(",")
        if line[6] == '':
            print("PAYOUT",bid,entry)
            c.execute("UPDATE bank_statement SET status='PAYOUT' WHERE bid=%s",(bid,))
        if line[5] == '':
            matches = re.findall("[A-Z0-9]{8}",line[4])
            if len(matches) >= 1:
                good_reference = False
                for match in matches:
                    deposit_reference = match.strip('., \t\r\n"')
                    amount = int(decimal.Decimal(line[6]) * ( 10 ** 8))
                    c.execute('SELECT uid FROM users WHERE deposref=%s',(deposit_reference,))
                    result = c.fetchone()
                    if result:
                        uid = result[0]
                        
                        c.execute("""
                        UPDATE
                            bank_statement
                        SET
                            status='PROC'
                        WHERE
                            bid=%s""",(bid,))
                            
                        c.execute("""
                        INSERT INTO requests (
                            req_type,
                            curr_type,
                            uid,
                            amount
                        )
                        VALUES
                        (
                            'DEPOS',
                            'GBP',
                            %s,
                            %s
                        )""",(uid,amount))
                        
                        reqid = c.lastrowid
                        
                        c.execute("""
                        UPDATE
                            bank_statement
                        SET
                            reqid=%s,
                            status='FINAL'
                        WHERE
                            bid=%s""",(reqid,bid))
                            
                        print("DEPOS",bid,uid,amount,reqid,entry)
                        
                        good_reference = True
                        break
                        
                if not good_reference:
                    print("BADREF",entry)
                    c.execute("UPDATE bank_statement SET status='BADREF' WHERE bid=%s",(bid,))
    except StopIteration:
        pass

